### PR TITLE
adapter: Prevent resource leaks in connection DDL

### LINKED ISSUE #16072: PrivateLink limit verification happens too late
### What version of Materialize are you using?

v0.31.0-dev

### How did you install Materialize?

Materialize Cloud

### What is the issue?

Verification of `max_aws_privatelink_connections` happens too late. If the `CREATE CONNECTION` fails with a "limit exceeded" error, while the connection object is not installed in the catalog, the `VpcEndpoint` resource _is_. This means we actually try to initiate a connection to the PrivateLink endpoint... and because we don't have GC of these `VpcEndpoint` resources, they will never get cleaned up.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by jkosh44**
For certain connections, such as SSH and AWS PrivateLink, an external resource was created before `catalog_transact` was run. If `catalog_transact` failed, then those resources would be leaked forever.

This commit updates the connection DDL so that the external resources are only created if `catalog_transact` was successful.

Fixes #16072

### Motivation
This PR fixes a recognized bug.

### Tips for reviewer
I had to make the `catalog_transact` callback async, which touched a lot of code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

**Comment by benesch** [comment]
Excellent, thank you! Can you update the PrivateLink cloudtest to have a not_exists assertion, too, when the limit is hit? I think the existing test is _already_ leaking a connection, and we just didn't test for it.

**Comment by benesch** [comment]
A few thoughts, based on our [Slack conversation](https://materializeinc.slack.com/archives/C02FWJ94HME/p1668640583709659?thread_ts=1668640402.973659&cid=C02FWJ94HME)!

* VPC endpoints and SSH keys are fundamentally different in severity. If we tell the user "VPC endpoint creation failed", we definitely do not want to be creating a VPC endpoint behind the scenes, as that results in observably wrong behavior: we'll actually try to initiate a connection to their endpoint service. And they won't have any way to tell Materialize to stop, since that VPC endpoint is not attached to a real connection! Whereas orphaning an SSH key is no problem; users won't be able to see it; and so we can GC it at our leisure.
* It's probably time to bite the bullet and GC orphaned secrets (https://github.com/MaterializeInc/materialize/issues/13579) and VPC endpoints. We already have a few orphaned VPC endpoints in production (due to people typing `CREATE CONNECTION` when their limit was 0), and I'll feel better if we get those cleaned up on the next release via a GC job. This is not wholly straightforward, but I think the strategy I just proposed for GC'ing orphaned compute clusters and storage hosts would work here too. See https://github.com/MaterializeInc/materialize/pull/16114#pullrequestreview-1183751981.

File: src/adapter/src/coord/sequencer.rs [hunk 1]
```diff
@@ -1320,18 +1300,22 @@ impl<S: Append + 'static> Coordinator<S> {
...
         let result = self
-            .catalog_transact(Some(&session), ops, move |txn| async move {
+            .catalog_transact_with(Some(&session), ops, move |txn| {
                 // Validate that the from collection is in fact a persist collection we can export.
-                txn.dataflow_client
-                    .storage
-                    .collection(sink.from)
-                    .map_err(|e| match e {
-                        StorageError::IdentifierMissing(_) => AdapterError::Unstructured(anyhow!(
+                std::future::ready(
```
Comment (petrosagg) [inline src/adapter/src/coord/sequencer.rs:1305]: @jkosh44 this is the trick ^. The closure no longer returns an `async { .. }` block. It does the computation immediately and returns a `Ready` future that contains the result directly

**Review by petrosagg — APPROVED** [review]

**Review by benesch — APPROVED** [review]

**Comment by benesch** [comment]
Thank you!!

**Pull request merged**